libbpf-tools/cpufreq.bpf.c

```diff
@@ -17,11 +17,21 @@ struct {
 	__type(value, struct hist);
 } hists SEC(".maps");
 
+#define clamp_umax(VAR, UMAX)						\
+	asm volatile (							\
+		"if %0 <= %[max] goto +1\n"				\
+		"%0 = %[max]\n"						\
+		: "+r"(VAR)						\
+		: [max]"i"(UMAX)					\
+	)
+
 SEC("tp_btf/cpu_frequency")
 int BPF_PROG(cpu_frequency, unsigned int state, unsigned int cpu_id)
 {
 	if (cpu_id >= MAX_CPU_NR)
 		return 0;
+
+	clamp_umax(cpu_id, MAX_CPU_NR - 1);
```

Comment:
I can't repro this locally. Could you please try:

```c
freqs_mhz[cpu_id & (MAX_CPU_NR - 1)] = state / 1000;
```

Reply:
It's probably Clang version differences. Try building recent Clang from sources, maybe? Anding with MAX_CPU_NR - 1 was the first thing I did, it didn't help, unfortunately, which made go with this BPF assembly helper approach.

Reply:
@yonghong-song FYI, I've tried `cpu_id & (MAX_CPU_NR - 1)` and did `barrier_var()`, but it didn't help. Could be some other corner case in Clang. The problem is the usual copying of registers, checking one of them, but then using another for indexing into the array.

Reply:
I am using latest clang, it seems the problem is fixed.
```
[yhs@devbig309.ftw3 ~/work/bcc/libbpf-tools] sudo bpftool prog dump xlated id 59580
int cpu_frequency(unsigned long long * ctx):
; int BPF_PROG(cpu_frequency, unsigned int state, unsigned int cpu_id)
   0: (79) r2 = *(u64 *)(r1 +8)
   1: (bf) r3 = r2
   2: (67) r3 <<= 32
   3: (77) r3 >>= 32
; if (cpu_id >= MAX_CPU_NR)
   4: (25) if r3 > 0x7f goto pc+13
; int BPF_PROG(cpu_frequency, unsigned int state, unsigned int cpu_id)
   5: (79) r1 = *(u64 *)(r1 +0)
; clamp_umax(cpu_id, MAX_CPU_NR - 1);
   6: (b5) if r2 <= 0x7f goto pc+1
   7: (b7) r2 = 127
; freqs_mhz[cpu_id] = state / 1000;
   8: (67) r2 <<= 32
   9: (77) r2 >>= 32
  10: (67) r2 <<= 2
  11: (18) r3 = map[id:31804][0]+0
  13: (0f) r3 += r2
; int BPF_PROG(cpu_frequency, unsigned int state, unsigned int cpu_id)
  14: (67) r1 <<= 32
  15: (77) r1 >>= 32
; freqs_mhz[cpu_id] = state / 1000;
  16: (37) r1 /= 1000
; freqs_mhz[cpu_id] = state / 1000;
  17: (63) *(u32 *)(r3 +0) = r1
; int BPF_PROG(cpu_frequency, unsigned int state, unsigned int cpu_id)
  18: (b7) r0 = 0
  19: (95) exit
```

So this probably an old clang issue. Could you do a double check?

Reply:
@anakryiko I just pushed a commit to sync with latest libbpf. You can rebase on top of the latest master branch.

Reply:
@yonghong-song, you are right, I had Clang from Nov 19th and it wasn't new enough. Pulling the very latest and rebuilding fixes the issue.

I think leaving clamp_umax() is still good and will make libbpf-tools buildable by wider variety of Clang versions, but let me know if you think I should drop that patch.

Reply:
actually, my bad, I haven't updated master properly to undo my fix. Re-doing everything, made sure I have latest Clang built. I have BCC state right before my patches, built everything and ran cpufreq. It is still broken.

```
$ clang --version
clang version 14.0.0 (https://github.com/llvm/llvm-project.git de904900600f11a65cdf44023061600b668e9df5)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/andriin/bin

$ sudo ./cpufreq
libbpf: prog 'cpu_frequency': BPF program load failed: Permission denied
libbpf: prog 'cpu_frequency': -- BEGIN PROG LOAD LOG --
R1 type=ctx expected=fp
; int BPF_PROG(cpu_frequency, unsigned int state, unsigned int cpu_id)
0: (79) r2 = *(u64 *)(r1 +8)
1: (18) r3 = 0xffffff80
; if (cpu_id >= MAX_CPU_NR)
3: (bf) r4 = r2
4: (5f) r4 &= r3
; if (cpu_id >= MAX_CPU_NR)
5: (55) if r4 != 0x0 goto pc+11
 R1=ctx(id=0,off=0,imm=0) R2_w=inv(id=1) R3_w=inv4294967168 R4_w=inv0 R10=fp0
;
6: (67) r2 <<= 32
7: (77) r2 >>= 32
; freqs_mhz[cpu_id] = state / 1000;
8: (67) r2 <<= 2
9: (18) r3 = 0xffffc9000ceaa000
11: (0f) r3 += r2
last_idx 11 first_idx 0
regs=4 stack=0 before 9: (18) r3 = 0xffffc9000ceaa000
regs=4 stack=0 before 8: (67) r2 <<= 2
regs=4 stack=0 before 7: (77) r2 >>= 32
regs=4 stack=0 before 6: (67) r2 <<= 32
regs=4 stack=0 before 5: (55) if r4 != 0x0 goto pc+11
regs=4 stack=0 before 4: (5f) r4 &= r3
regs=4 stack=0 before 3: (bf) r4 = r2
regs=4 stack=0 before 1: (18) r3 = 0xffffff80
regs=4 stack=0 before 0: (79) r2 = *(u64 *)(r1 +8)
; int BPF_PROG(cpu_frequency, unsigned int state, unsigned int cpu_id)
12: (79) r1 = *(u64 *)(r1 +0)
13: (67) r1 <<= 32
14: (77) r1 >>= 32
; freqs_mhz[cpu_id] = state / 1000;
15: (37) r1 /= 1000
; freqs_mhz[cpu_id] = state / 1000;
16: (63) *(u32 *)(r3 +0) = r1
 R1_w=inv(id=0) R2_w=invP(id=0,umax_value=17179869180,var_off=(0x0; 0x3fffffffc),s32_max_value=2147483644,u32_max_value=-4) R3_w=map_value(id=0,off=0,ks=4,vs=720,umax_value=17179869180,var_off=(0x0; 0x3fffffffc),s32_max_value=2147483644,u32_max_value=-4) R4_w=inv0 R10=fp0
R3 unbounded memory access, make sure to bounds check any such access
processed 15 insns (limit 1000000) max_states_per_insn 0 total_states 0 peak_states 0 mark_read 0
-- END PROG LOAD LOG --
libbpf: failed to load program 'cpu_frequency'
libbpf: failed to load object 'cpufreq_bpf'
libbpf: failed to load BPF skeleton 'cpufreq_bpf': -13
failed to open and/or load BPF object
```